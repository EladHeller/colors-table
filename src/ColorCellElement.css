#container {
  width: 50px;
  height: 50px;
}

#container:not(.primary) {
  cursor: pointer;
}