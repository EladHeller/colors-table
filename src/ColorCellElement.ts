import { attr, controller } from '@github/catalyst';
import style from './ColorCellElement.css';

@controller
export default class ColorCellElement extends HTMLElement {
  @attr cellColor = 'black';

  @attr isActive = false;

  @attr isPrimary = false;

  private container = document.createElement('div');

  constructor() {
    super();
    const shadow = this.attachShadow({ mode: 'open' });
    const styleElement = document.createElement('style');
    styleElement.textContent = style.toString();
    shadow.appendChild(styleElement);
    shadow.appendChild(this.container);
    this.container.id = 'container';
  }

  connectedCallback() {
    this.render();
  }

  attributeChangedCallback() {
    this.render();
  }

  render() {
    this.container.style.border = `${this.cellColor} 5px solid`;
    if (this.isActive) {
      this.container.style.backgroundColor = this.cellColor;
    }
    if (this.isPrimary) {
      this.container.classList.toggle('primary', this.isPrimary);
    }
  }
}
