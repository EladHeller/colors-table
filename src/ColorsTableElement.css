table, th, td {
  border: 2px solid #c2c2ce;
  border-collapse: collapse;
}

th, td {
  padding: 5px;
}

td:nth-child(2), th:nth-child(2) {
  white-space: nowrap;
      width: 1px;

}

.error {
  background-color: #f7d5d5;
  color: red;
  font-weight: bold;
}


.color-cell {
  display: inline-block;
  margin-left: 10px;
  width: 50px;
  height: 50px;
}

.color-cell:not(.primary) {
  cursor: pointer;
}