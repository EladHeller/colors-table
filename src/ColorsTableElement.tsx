import { createRoot, Root } from 'react-dom/client';
import { attr, controller } from '@github/catalyst';
import { fromEvent, map } from 'rxjs';
import { IColorRow, IColorsView } from './types';
import style from './ColorsTableElement.css';

type ColorClickedCallback = (color: string) => void;

interface IColorCellProps {
  cellColor: string;
  isActive: boolean;
  isPrimary?: boolean;
  onClick?: () => void;
}

function ColorCellElement(props: IColorCellProps) {
  const {
    cellColor, isActive, isPrimary, onClick,
  } = props;
  const styleData = {
    backgroundColor: isActive ? cellColor : '',
    border: `${cellColor} 5px solid`,
  };
  return (
    <div
      className={`color-cell ${isActive ? 'active' : ''} ${isPrimary ? 'primary' : ''}`}
      style={styleData}
      onClick={onClick}
    />
  );
}

interface IColorsTableProps {
  colors: IColorRow[];
  onColorClick: ColorClickedCallback;
  isError: boolean;
}

function ColorsTableBody(props: IColorsTableProps) {
  const { colors, onColorClick, isError } = props;
  if (isError) {
    return (
        <tr className='error'>
          <td colSpan={3}>There was an error while loading the colors.</td>
        </tr>
    );
  }
  if (colors.length) {
    return (
      <>
        {colors.map((row, i) => (
          <tr key={i}>
            <td>{i + 1}</td>
            <td>
              {row.cells.map((cell, index) => (
                <ColorCellElement
                  key={index}
                  cellColor={cell.color}
                  isActive={cell.isActive}
                  onClick={() => onColorClick(cell.color)}
                />
              ))}
            </td>
            <td>
              <ColorCellElement cellColor={row.primaryColor} isActive isPrimary />
            </td>
          </tr>
        ))}
      </>
    );
  }
  return (
      <tr>
        <td colSpan={3}>loading colors...</td>
      </tr>
  );
}

export function ColorsTable(props: IColorsTableProps) {
  const { colors, onColorClick, isError } = props;

  return (
    <table>
      <thead>
        <tr>
          <th>#</th>
          <th>Colors</th>
          <th>Primary</th>
        </tr>
      </thead>
      <tbody>
      <ColorsTableBody colors={colors} isError={isError} onColorClick={onColorClick}/>
      </tbody>
    </table>
  );
}

@controller
export default class ColorsTableElement extends HTMLElement implements IColorsView {
  private colorTable: IColorRow[] = [];

  private root: Root;

  private tableContainer = document.createElement('div');

  @attr isErrored = false;

  colorClicked$ = fromEvent(this, 'color-clicked')
    .pipe(
      map((event: Event) => (event as CustomEvent).detail),
    );

  set colors(colors: IColorRow[]) {
    this.colorTable = colors;
    this.render();
  }

  get colors(): IColorRow[] {
    return this.colorTable;
  }

  constructor() {
    super();
    const shadow = this.attachShadow({ mode: 'open' });
    shadow.appendChild(this.tableContainer);
    const styleElement = document.createElement('style');
    styleElement.textContent = style.toString();
    shadow.appendChild(styleElement);
    this.root = createRoot(this.tableContainer);
    this.root.render(<ColorsTable
      colors={this.colors}
      isError={this.isErrored}
      onColorClick={(color) => this.onCellClick(color)}
    />);
  }

  private onCellClick(color: string) {
    this.dispatchEvent(new CustomEvent('color-clicked', { detail: color }));
  }

  private render() {
    this.root.render(<ColorsTable
      colors={this.colors}
      isError={this.isErrored}
      onColorClick={(color) => this.onCellClick(color)}
    />);
  }

  connectedCallback() {
    this.render();
  }

  attributeChangedCallback() {
    this.render();
  }
}
