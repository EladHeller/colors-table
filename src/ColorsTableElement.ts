import { controller } from '@github/catalyst';
import { html, render } from '@github/jtml';
import { fromEvent, map } from 'rxjs';
import { IColorRow, IColorsView } from './types';
import { clearChildren } from './utilities';
import style from './ColorsTableElement.css';
import ColorCellElement from './ColorCellElement';

type ColorClickedCallback = (color: string) => void;

function createColorCell(
  color: string,
  isActive: boolean,
  isPrimary: boolean,
  onClick?: ColorClickedCallback,
) {
  const cellElement = new ColorCellElement();
  cellElement.cellCollor = color;
  cellElement.isActive = isPrimary || isActive;
  cellElement.isPrimary = isPrimary;
  cellElement.addEventListener('click', () => {
    onClick?.(color);
  });
  return cellElement;
}

@controller
export default class ColorsTableElement extends HTMLElement implements IColorsView {
  private tableBody = document.createElement('tbody');

  private colorTable: IColorRow[] = [];

  colorClicked$ = fromEvent(this, 'color-clicked')
    .pipe(
      map((event: Event) => (event as CustomEvent).detail),
    );

  set colors(colors: IColorRow[]) {
    this.colorTable = colors;
    this.renderColorsTable();
  }

  get colors(): IColorRow[] {
    return this.colorTable;
  }

  constructor() {
    super();
    const shadow = this.attachShadow({ mode: 'open' });
    const table = document.createElement('table');
    shadow.appendChild(table);
    render(html`<thead><tr><th>#</th><th>Colors</th><th>Primary</th></tr></thead>`, table);
    table.appendChild(this.tableBody);

    const styleElement = document.createElement('style');
    styleElement.textContent = style;
    shadow.appendChild(styleElement);
  }

  private onCellClick(color: string) {
    this.dispatchEvent(new CustomEvent('color-clicked', { detail: color }));
  }

  private renderColorsTable() {
    clearChildren(this.tableBody);
    this.colors.forEach((row, i) => {
      const rowElement = document.createElement('tr');
      this.tableBody.appendChild(rowElement);

      const numberCell = document.createElement('td');
      const colorCell = document.createElement('td');
      const primaryColor = document.createElement('td');

      numberCell.textContent = (i + 1).toString();

      row.cells.forEach((cell) => {
        colorCell.appendChild(createColorCell(cell.color, cell.isActive, false, () => {
          this.onCellClick(cell.color);
        }));
      });

      primaryColor.appendChild(createColorCell(row.primaryColor, true, true));

      rowElement.appendChild(numberCell);
      rowElement.appendChild(colorCell);
      rowElement.appendChild(primaryColor);
    });
  }
}
