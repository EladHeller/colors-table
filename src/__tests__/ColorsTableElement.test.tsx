import React from 'react';
import renderer from 'react-test-renderer';
import ColorCellElement from '../ColorCellElement';
import ColorsTableElement, { ColorsTable } from '../ColorsTableElement';

describe('ColorsTableElement', () => {
  let element: ColorsTableElement;

  beforeEach(() => {
    element = new ColorsTableElement();
    document.body.appendChild(element);
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('should correctly get and set the colors property', () => {
    expect(element.colors).toEqual([]);

    const colors = [{ cells: [{ color: 'yellow', isActive: true }], primaryColor: 'yellow' },
      { cells: [{ color: 'red', isActive: false }], primaryColor: 'red' },
    ];
    element.colors = colors;
    expect(element.colors).toEqual(colors);
  });

  it('should render the table with the correct number of rows and cells', () => {
    const colors = [{ cells: [{ color: 'yellow', isActive: true }], primaryColor: 'yellow' },
      { cells: [{ color: 'red', isActive: false }], primaryColor: 'red' },
    ];
    const component = renderer.create(<ColorsTable
      colors={colors} isError={false} onColorClick={jest.fn()}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('should dispatch the correct event when a color cell is clicked', () => {
  //   const colors = [{ cells: [{ color: 'yellow', isActive: true }], primaryColor: 'yellow' },
  //     { cells: [{ color: 'red', isActive: false }], primaryColor: 'red' },
  //   ];
  //   const onColorClick = jest.fn();
  //   const component = renderer.create(<ColorsTable
  //     colors={colors} isError={false} onColorClick={onColorClick}
  //   />);
  //   renderer.act(() => {
  //     tree.props.onColorClick();
  //   });

  //   const tree = component.toJSON();
  //   expect(onColorClick).toBeCalled();
  // });
  //   let colorFromClick = '';

  //   element.colors = [
  //     { cells: [{ color: 'yellow', isActive: true }, { color: 'red', isActive: true }], primaryColor: 'yellow' },
  //     { cells: [{ color: 'red', isActive: false }], primaryColor: 'red' },
  //   ];

  //   element.colorClicked$.subscribe((color) => {
  //     colorFromClick = color;
  //   });

  //   const shadowRoot = element.shadowRoot as ShadowRoot;
  //   const tableBody = shadowRoot.querySelector('tbody');
  //   const firstRow = tableBody?.children[0] as HTMLTableRowElement;
  //   const firstCell = firstRow.children[1].children[0] as ColorCellElement;
  //   firstCell.click();

  //   expect(colorFromClick).toBe('yellow');
  // });

  it('should render error message if isErrored is true', () => {
    const component = renderer.create(<ColorsTable
      colors={[]} isError
      onColorClick={jest.fn()}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('should render error message even if it creates with isError', () => {
  //   document.body.innerHTML = '';
  //   element = new ColorsTableElement();
  //   element.isErrored = true;
  //   document.body.appendChild(element);

  //   const shadowRoot = element.shadowRoot as ShadowRoot;
  //   const errorMessage = shadowRoot.querySelector('td');

  //   expect(errorMessage?.textContent).toBe('There was an error loading the colors.');
  // });

  it('should render loading message if until there is data', () => {
    const component = renderer.create(<ColorsTable
      colors={[]} isError={false} onColorClick={jest.fn()}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
