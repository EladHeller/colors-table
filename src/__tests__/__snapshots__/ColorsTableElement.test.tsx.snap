// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorsTableElement should render error message if isErrored is true 1`] = `
<table>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        Colors
      </th>
      <th>
        Primary
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="error"
    >
      <td
        colSpan={3}
      >
        There was an error while loading the colors.
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ColorsTableElement should render loading message if until there is data 1`] = `
<table>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        Colors
      </th>
      <th>
        Primary
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={3}
      >
        loading colors...
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ColorsTableElement should render the table with the correct number of rows and cells 1`] = `
<table>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        Colors
      </th>
      <th>
        Primary
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        <div
          className="color-cell active "
          onClick={[Function]}
          style={
            {
              "backgroundColor": "yellow",
              "border": "yellow 5px solid",
            }
          }
        />
      </td>
      <td>
        <div
          className="color-cell active primary"
          style={
            {
              "backgroundColor": "yellow",
              "border": "yellow 5px solid",
            }
          }
        />
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        <div
          className="color-cell  "
          onClick={[Function]}
          style={
            {
              "backgroundColor": "",
              "border": "red 5px solid",
            }
          }
        />
      </td>
      <td>
        <div
          className="color-cell active primary"
          style={
            {
              "backgroundColor": "red",
              "border": "red 5px solid",
            }
          }
        />
      </td>
    </tr>
  </tbody>
</table>
`;
